source/extensions/filters/http/oauth2/filter.h

```diff
@@ -0,0 +1,232 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "envoy/common/matchers.h"
+#include "envoy/config/core/v3/http_uri.pb.h"
+#include "envoy/extensions/filters/http/oauth2/v3alpha/oauth.pb.h"
+#include "envoy/http/header_map.h"
+#include "envoy/server/filter_config.h"
+#include "envoy/stats/stats_macros.h"
+#include "envoy/stream_info/stream_info.h"
+#include "envoy/upstream/cluster_manager.h"
+
+#include "common/common/assert.h"
+#include "common/common/matchers.h"
+#include "common/config/datasource.h"
+#include "common/formatter/substitution_formatter.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/header_utility.h"
+#include "common/http/rest_api_fetcher.h"
+#include "common/http/utility.h"
+
+#include "extensions/filters/http/common/pass_through_filter.h"
+#include "extensions/filters/http/oauth2/filter.h"
+#include "extensions/filters/http/oauth2/oauth.h"
+#include "extensions/filters/http/oauth2/oauth_client.h"
+
+#include "absl/strings/str_split.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Oauth2 {
+
+class OAuth2Client;
+
+// Helper class used to fetch secrets (usually from SDS).
+class SecretReader {
+public:
+  virtual ~SecretReader() = default;
+  virtual std::string clientSecret() const PURE;
+  virtual std::string tokenSecret() const PURE;
+};
+
+class SDSSecretReader : public SecretReader {
+public:
+  SDSSecretReader(Secret::GenericSecretConfigProviderSharedPtr client_secret_provider,
+                  Secret::GenericSecretConfigProviderSharedPtr token_secret_provider, Api::Api& api)
+      : client_secret_provider_(std::move(client_secret_provider)),
+        token_secret_provider_(std::move(token_secret_provider)), api_(api) {}
+
+  std::string clientSecret() const override {
+    const auto* secret = client_secret_provider_->secret();
+    if (!secret) {
+      return EMPTY_STRING;
+    }
+
+    return Config::DataSource::read(secret->secret(), true, api_);
```

Comment:
From the call pattern this would read disk on every request. Cache it in memory and update in callback from secret provider. Same below.